linking issue on trusty with and latest nightly

linking of application using heapsize and built with latest nightly fails on ubuntu trusty

example:

https://github.com/debris/heapsize_issue

travis build matrix:

https://travis-ci.org/debris/heapsize_issue

```
error: linking with `cc` failed: exit code: 1

note: "cc" "-Wl,--as-needed" "-Wl,-z,noexecstack" "-m64" "-L" "/home/travis/rust/lib/rustlib/x86_64-unknown-linux-gnu/lib" "/home/travis/build/debris/heapsize_issue/target/debug/app.0.o" "-o" "/home/travis/build/debris/heapsize_issue/target/debug/app" "-Wl,--gc-sections" "-pie" "-nodefaultlibs" "-L" "/home/travis/build/debris/heapsize_issue/target/debug" "-L" "/home/travis/build/debris/heapsize_issue/target/debug/deps" "-L" "/home/travis/rust/lib/rustlib/x86_64-unknown-linux-gnu/lib" "-Wl,-Bstatic" "-Wl,-Bdynamic" "/home/travis/build/debris/heapsize_issue/target/debug/libheapsize_issue.rlib" "/home/travis/build/debris/heapsize_issue/target/debug/deps/libheapsize-a00ec68b9e8bd80c.rlib" "/home/travis/rust/lib/rustlib/x86_64-unknown-linux-gnu/lib/libstd-fd663c41.rlib" "/home/travis/rust/lib/rustlib/x86_64-unknown-linux-gnu/lib/libcollections-fd663c41.rlib" "/home/travis/rust/lib/rustlib/x86_64-unknown-linux-gnu/lib/librustc_unicode-fd663c41.rlib" "/home/travis/rust/lib/rustlib/x86_64-unknown-linux-gnu/lib/librand-fd663c41.rlib" "/home/travis/rust/lib/rustlib/x86_64-unknown-linux-gnu/lib/liballoc-fd663c41.rlib" "/home/travis/rust/lib/rustlib/x86_64-unknown-linux-gnu/lib/liballoc_jemalloc-fd663c41.rlib" "/home/travis/rust/lib/rustlib/x86_64-unknown-linux-gnu/lib/liblibc-fd663c41.rlib" "/home/travis/rust/lib/rustlib/x86_64-unknown-linux-gnu/lib/libcore-fd663c41.rlib" "-l" "dl" "-l" "pthread" "-l" "gcc_s" "-l" "pthread" "-l" "c" "-l" "m" "-l" "rt" "-l" "compiler-rt"

note: /home/travis/build/debris/heapsize_issue/target/debug/deps/libheapsize-a00ec68b9e8bd80c.rlib(heapsize-a00ec68b9e8bd80c.0.o): In function `heapsize::heap_size_of':

/home/travis/.cargo/registry/src/github.com-88ac128001ac3a9a/heapsize-0.3.2/src/lib.rs:45: undefined reference to `je_malloc_usable_size'

collect2: error: ld returned 1 exit status
```
@larsbergstrom is this related to other issues we've seen recently?

--- a/.travis.yml
+++ b/.travis.yml
@@ -1,10 +1,18 @@
 language: rust
-rust: nightly
+rust:
+ # Jemalloc prefixing change between these two versions
+ - nightly-2016-02-14
+ - nightly-2016-02-17
+
+ - nightly
+ - beta
+ - stable
+
 notifications:
   webhooks: http://build.servo.org:54856/travis
 
 script:
  - cargo test
- - "[ $TRAVIS_RUST_VERSION = nightly ] || cargo test --features unstable"
- - "[ $TRAVIS_RUST_VERSION = nightly ] || cargo test --manifest-path plugin/Cargo.toml"
+ - "[ $TRAVIS_RUST_VERSION != nightly ] || cargo test --features unstable"
+ - "[ $TRAVIS_RUST_VERSION != nightly ] || cargo test --manifest-path plugin/Cargo.toml"
 

--- a/Cargo.toml
+++ b/Cargo.toml
@@ -8,8 +8,7 @@ repository = "https://github.com/servo/heapsize"
 build = "build.rs"
 
 [build-dependencies]
-semver = "0.2"
-chrono = "0.2"
+regex = "0.1"
 
 [features]
 unstable = []

--- a/Cargo.toml
+++ b/Cargo.toml
@@ -8,3 +8,4 @@ repository = "https://github.com/servo/heapsize"
 
 [features]
 unstable = []
+nightly = []

--- a/Cargo.toml
+++ b/Cargo.toml
@@ -5,7 +5,11 @@ authors = [ "The Servo Project Developers" ]
 description = "Infrastructure for measuring the total runtime size of an object on the heap"
 license = "MPL-2.0"
 repository = "https://github.com/servo/heapsize"
+build = "build.rs"
+
+[build-dependencies]
+semver = "0.2"
+chrono = "0.2"
 
 [features]
 unstable = []
-nightly = []

--- a/build.rs
+++ b/build.rs
@@ -1,0 +1,31 @@
+extern crate semver;
+extern crate chrono;
+
+use std::env::var;
+use std::process::Command;
+use semver::Version;
+use chrono::NaiveDate;
+
+fn main() {
+    let unprefixed_jemalloc_version: Version = "1.8.0-dev".parse().unwrap();
+    let unprefixed_jemalloc_day: NaiveDate = "2016-02-16".parse().unwrap();
+
+    let rustc_version_string = Command::new(var("RUSTC").unwrap_or("rustc".into()))
+        .arg("--version").output().ok()
+        .and_then(|o| String::from_utf8(o.stdout).ok())
+        .unwrap();
+    let rustc_version: Version = rustc_version_string.split_whitespace()
+        .skip(1).next()
+        .and_then(|r| r.parse().ok())
+        .unwrap();
+    let rustc_day: NaiveDate = rustc_version_string.split_whitespace()
+        .last()
+        .and_then(|r| r[..r.len()-1].parse().ok())
+        .unwrap();
+
+    if rustc_day < unprefixed_jemalloc_day
+        || rustc_version < unprefixed_jemalloc_version
+    {
+        println!("cargo:rustc-cfg=prefixed_jemalloc");
+    }
+}

--- a/build.rs
+++ b/build.rs
@@ -1,31 +1,26 @@
-extern crate semver;
-extern crate chrono;
+extern crate regex;
 
 use std::env::var;
 use std::process::Command;
-use semver::Version;
-use chrono::NaiveDate;
+use std::str;
 
 fn main() {
-    let unprefixed_jemalloc_version: Version = "1.8.0-dev".parse().unwrap();
-    let unprefixed_jemalloc_day: NaiveDate = "2016-02-16".parse().unwrap();
-
-    let rustc_version_string = Command::new(var("RUSTC").unwrap_or("rustc".into()))
-        .arg("--version").output().ok()
-        .and_then(|o| String::from_utf8(o.stdout).ok())
+    let version_line = Command::new(var("RUSTC").unwrap_or("rustc".into()))
+        .arg("--version")
+        .output()
+        .unwrap()
+        .stdout;
+    let captures = regex::Regex::new(r"rustc (\d+)\.(\d+)\.(\d+).+(\d{4}-\d{2}-\d{2})\)")
+        .unwrap()
+        .captures(str::from_utf8(&version_line).unwrap())
         .unwrap();
-    let rustc_version: Version = rustc_version_string.split_whitespace()
-        .skip(1).next()
-        .and_then(|r| r.parse().ok())
-        .unwrap();
-    let rustc_day: NaiveDate = rustc_version_string.split_whitespace()
-        .last()
-        .and_then(|r| r[..r.len()-1].parse().ok())
-        .unwrap();
-
-    if rustc_day < unprefixed_jemalloc_day
-        || rustc_version < unprefixed_jemalloc_version
-    {
+    let version = (
+        captures[1].parse::<u32>().unwrap(),
+        captures[2].parse::<u32>().unwrap(),
+        captures[3].parse::<u32>().unwrap(),
+        &captures[4],
+    );
+    if version < (1, 8, 0, "2016-02-14") {
         println!("cargo:rustc-cfg=prefixed_jemalloc");
     }
 }

--- a/src/lib.rs
+++ b/src/lib.rs
@@ -20,7 +20,7 @@ use std::rc::Rc;
 
 #[cfg(not(target_os = "windows"))]
 extern {
-    #![cfg_attr(feature = "nightly", link(name = "malloc_usable_size"))]
+    #[cfg_attr(any(prefixed_jemalloc, target_os = "macos", target_os = "android"), link_name = "je_malloc_usable_size")]
     // Get the size of a heap block.
     //
     // Ideally Rust would expose a function like this in std::rt::heap, which would avoid the
@@ -30,7 +30,7 @@ extern {
     // platforms `JEMALLOC_USABLE_SIZE_CONST` is `const` and on some it is empty. But in practice
     // this function doesn't modify the contents of the block that `ptr` points to, so we use
     // `*const c_void` here.
-    fn je_malloc_usable_size(ptr: *const c_void) -> usize;
+    fn malloc_usable_size(ptr: *const c_void) -> usize;
 }
 
 /// A wrapper for je_malloc_usable_size that handles `EMPTY` and returns `usize`.
@@ -43,7 +43,7 @@ pub unsafe fn heap_size_of(ptr: *const c_void) -> usize {
     if ptr == 0x01 as *const c_void {
         0
     } else {
-        je_malloc_usable_size(ptr)
+        malloc_usable_size(ptr)
     }
 }
 

--- a/src/lib.rs
+++ b/src/lib.rs
@@ -20,6 +20,7 @@ use std::rc::Rc;
 
 #[cfg(not(target_os = "windows"))]
 extern {
+    #![cfg_attr(feature = "nightly", link(name = "malloc_usable_size"))]
     // Get the size of a heap block.
     //
     // Ideally Rust would expose a function like this in std::rt::heap, which would avoid the

